File: backend/src/fined/call_analytics.py
"""Privacy-safe call outcomes and a public aggregate dashboard snapshot."""

from __future__ import annotations

import fcntl
import json
import os
import re
import secrets
import sqlite3
import tempfile
import time
from collections.abc import Callable, Iterator
from contextlib import contextmanager
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path

CALL_SUCCESS_DEFINITION = (
    "The learner completed a verified action: received grounded market evidence, "
    "received a verified tax rule, received trusted market or historical data, "
    "completed a paper fill or created a human-help request."
)
SUCCESS_CONDITIONS = frozenset(
    {
        "grounded_answer_delivered",
        "tax_rule_delivered",
        "market_quote_delivered",
        "historical_return_calculated",
        "paper_fill_completed",
        "human_help_created",
    }
)
FAILURE_TYPES = frozenset(
    {
        "no_completed_action",
        "incomplete",
        "no_response",
        "tool_unavailable",
        "system_error",
    }
)
CHANNELS = frozenset({"browser", "sip"})
_CALL_ID = re.compile(r"^CALL-(?:[A-F0-9]{4}-){5}[A-F0-9]{4}$")
_AGENT_STATES = frozenset({"initializing", "idle", "listening", "thinking", "speaking"})


class CallAnalyticsValidationError(ValueError):
    """Raised before arbitrary or identifying values can enter analytics."""


class AgentTalkTimeTracker:
    """Measure only validated LiveKit agent-speaking state intervals."""

    def __init__(self, *, clock: Callable[[], float] = time.monotonic) -> None:
        self._clock = clock
        self._totals = {"fined": 0.0, "taxed": 0.0}
        self._open_interval: tuple[str, float] | None = None
        self._closed = False

    def on_agent_state_changed(
        self, old_state: str, new_state: str, active_agent_name: object
    ) -> None:
        if self._closed:
            return
        if old_state not in _AGENT_STATES or new_state not in _AGENT_STATES:
            self._open_interval = None
            return
        if old_state == new_state:
            self._open_interval = None
            return
        if new_state == "speaking":
            if old_state == "speaking" or self._open_interval is not None:
                self._open_interval = None
                return
            if (
                not isinstance(active_agent_name, str)
                or active_agent_name not in self._totals
            ):
                return
            self._open_interval = (str(active_agent_name), self._clock())
            return
        if old_state == "speaking":
            self._close_open_interval()
            return
        if self._open_interval is not None:
            self._open_interval = None

    def close(self) -> dict[str, int]:
        if not self._closed:
            self._close_open_interval()
            self._closed = True
        return {
            "fined_talk_seconds": round(self._totals["fined"]),
            "taxed_talk_seconds": round(self._totals["taxed"]),
        }

    def _close_open_interval(self) -> None:
        interval = self._open_interval
        self._open_interval = None
        if interval is None:
            return
        agent_name, started_at = interval
        ended_at = self._clock()
        if ended_at < started_at:
            return
        self._totals[agent_name] += ended_at - started_at


def new_call_id() -> str:
    token = secrets.token_hex(12).upper()
    return "CALL-" + "-".join(token[index : index + 4] for index in range(0, 24, 4))


@dataclass(frozen=True)
class CallAnalyticsInput:
    call_id: str
    channel: str
    started_at: datetime
    ended_at: datetime
    success_condition: str | None = None
    failure_type: str | None = None
    fined_talk_seconds: int = 0
    taxed_talk_seconds: int = 0
    handoff_count: int = 0

    def __post_init__(self) -> None:
        if (
            not isinstance(self.call_id, str)
            or _CALL_ID.fullmatch(self.call_id) is None
        ):
            raise CallAnalyticsValidationError("call_id is invalid.")
        if self.channel not in CHANNELS:
            raise CallAnalyticsValidationError("channel is invalid.")
        if self.started_at.tzinfo is None or self.ended_at.tzinfo is None:
            raise CallAnalyticsValidationError("timestamps must include a timezone.")
        if self.ended_at < self.started_at:
            raise CallAnalyticsValidationError("ended_at cannot precede started_at.")
        if (
            self.success_condition is not None
            and self.success_condition not in SUCCESS_CONDITIONS
        ):
            raise CallAnalyticsValidationError("success_condition is invalid.")
        if self.failure_type is not None and self.failure_type not in FAILURE_TYPES:
            raise CallAnalyticsValidationError("failure_type is invalid.")
        if (self.success_condition is None) == (self.failure_type is None):
            raise CallAnalyticsValidationError(
                "Exactly one supported success or failure detail is required."
            )
        measurements = (
            self.fined_talk_seconds,
            self.taxed_talk_seconds,
            self.handoff_count,
        )
        if any(
            isinstance(value, bool) or not isinstance(value, int) or value < 0
            for value in measurements
        ):
            raise CallAnalyticsValidationError(
                "Agent measurements must be non-negative integers."
            )
        duration_seconds = round((self.ended_at - self.started_at).total_seconds())
        if self.fined_talk_seconds + self.taxed_talk_seconds > duration_seconds + 1:
            raise CallAnalyticsValidationError(
                "Agent speaking time cannot exceed call duration."
            )


class SQLiteCallAnalyticsStore:
    """Store minimal call outcomes and publish a caller-safe JSON snapshot."""

    def __init__(self, database_path: str | Path, *, snapshot_path: str | Path) -> None:
        self._database_path = Path(database_path)
        self._snapshot_path = Path(snapshot_path)
        self._snapshot_lock_path = self._snapshot_path.with_name(
            f".{self._snapshot_path.name}.lock"
        )
        self._database_path.parent.mkdir(parents=True, exist_ok=True)
        self._snapshot_path.parent.mkdir(parents=True, exist_ok=True)
        self._initialize()
        with self._publication_lock():
            self._write_snapshot(self.public_summary())

    def record(self, call: CallAnalyticsInput) -> None:
        if not isinstance(call, CallAnalyticsInput):
            raise CallAnalyticsValidationError("A call analytics record is required.")
        outcome = "successful" if call.success_condition is not None else "failed"
        detail = call.success_condition or call.failure_type
        duration_seconds = max(
            0, round((call.ended_at - call.started_at).total_seconds())
        )
        with self._publication_lock():
            with self._connect() as connection:
                connection.execute(
                    """
                    INSERT OR IGNORE INTO call_outcomes (
                        call_id, channel, started_at, ended_at, duration_seconds,
                        outcome, detail, fined_talk_seconds, taxed_talk_seconds,
                        handoff_count
                    ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                    """,
                    (
                        call.call_id,
                        call.channel,
                        call.started_at.isoformat(),
                        call.ended_at.isoformat(),
                        duration_seconds,
                        outcome,
                        detail,
                        call.fined_talk_seconds,
                        call.taxed_talk_seconds,
                        call.handoff_count,
                    ),
                )
            self._write_snapshot(self.public_summary())

    def public_summary(self) -> dict[str, object]:
        with self._connect() as connection:
            (
                total,
                successful,
                failed,
                total_duration,
                fined_talk,
                taxed_talk,
                handoffs,
            ) = connection.execute(
                """
                SELECT COUNT(*),
                       SUM(CASE WHEN outcome = 'successful' THEN 1 ELSE 0 END),
                       SUM(CASE WHEN outcome = 'failed' THEN 1 ELSE 0 END),
                       SUM(duration_seconds),
                       SUM(fined_talk_seconds),
                       SUM(taxed_talk_seconds),
                       SUM(handoff_count)
                FROM call_outcomes
                """
            ).fetchone()
            rows = connection.execute(
                """
                SELECT call_id, started_at, duration_seconds, channel, outcome, detail,
                       fined_talk_seconds, taxed_talk_seconds, handoff_count
                FROM call_outcomes
                ORDER BY started_at DESC, call_id DESC
                LIMIT 20
                """
            ).fetchall()
        total = int(total or 0)
        successful = int(successful or 0)
        failed = int(failed or 0)
        rate = round(successful * 100 / total, 1) if total else 0.0
        return {
            "version": 2,
            "success_definition": CALL_SUCCESS_DEFINITION,
            "totals": {
                "total_calls": total,
                "successful_calls": successful,
                "failed_calls": failed,
                "success_rate_percent": rate,
                "total_duration_seconds": int(total_duration or 0),
                "fined_talk_seconds": int(fined_talk or 0),
                "taxed_talk_seconds": int(taxed_talk or 0),
                "handoff_count": int(handoffs or 0),
            },
            "recent_calls": [
                {
                    "call_id": row[0],
                    "started_at": row[1],
                    "duration_seconds": row[2],
                    "channel": row[3],
                    "outcome": row[4],
                    "detail": row[5],
                    "fined_talk_seconds": row[6],
                    "taxed_talk_seconds": row[7],
                    "handoff_count": row[8],
                }
                for row in rows
            ],
        }

    def _connect(self) -> sqlite3.Connection:
        return sqlite3.connect(self._database_path, timeout=5)

    def _initialize(self) -> None:
        with self._connect() as connection:
            self._enable_write_ahead_log(connection)
            connection.execute("BEGIN IMMEDIATE")
            connection.execute(
                """
                CREATE TABLE IF NOT EXISTS call_outcomes (
                    call_id TEXT PRIMARY KEY,
                    channel TEXT NOT NULL,
                    started_at TEXT NOT NULL,
                    ended_at TEXT NOT NULL,
                    duration_seconds INTEGER NOT NULL,
                    outcome TEXT NOT NULL,
                    detail TEXT NOT NULL
                )
                """
            )
            columns = {
                row[1] for row in connection.execute("PRAGMA table_info(call_outcomes)")
            }
            for column in (
                "fined_talk_seconds",
                "taxed_talk_seconds",
                "handoff_count",
            ):
                if column not in columns:
                    connection.execute(
                        f"ALTER TABLE call_outcomes "
                        f"ADD COLUMN {column} INTEGER NOT NULL DEFAULT 0"
                    )

    @staticmethod
    def _enable_write_ahead_log(connection: sqlite3.Connection) -> None:
        retry_delays = (0.025, 0.05, 0.1, 0.2, 0.4)
        for retry_delay in (*retry_delays, None):
            try:
                connection.execute("PRAGMA journal_mode=WAL")
                return
            except sqlite3.OperationalError as error:
                if "locked" not in str(error).casefold() or retry_delay is None:
                    raise
                time.sleep(retry_delay)

    @contextmanager
    def _publication_lock(self) -> Iterator[None]:
        file_descriptor = os.open(
            self._snapshot_lock_path,
            os.O_CREAT | os.O_RDWR,
            0o600,
        )
        with os.fdopen(file_descriptor, "rb+") as lock_file:
            fcntl.flock(lock_file.fileno(), fcntl.LOCK_EX)
            try:
                yield
            finally:
                fcntl.flock(lock_file.fileno(), fcntl.LOCK_UN)

    def _write_snapshot(self, summary: dict[str, object]) -> None:
        payload = json.dumps(
            summary, ensure_ascii=False, separators=(",", ":"), sort_keys=True
        )
        file_descriptor, temporary_name = tempfile.mkstemp(
            dir=self._snapshot_path.parent,
            prefix=f".{self._snapshot_path.name}.",
            suffix=".tmp",
            text=True,
        )
        try:
            with os.fdopen(file_descriptor, "w", encoding="utf-8") as temporary:
                temporary.write(payload)
                temporary.flush()
                os.fsync(temporary.fileno())
            os.replace(temporary_name, self._snapshot_path)
        finally:
            if os.path.exists(temporary_name):
                os.unlink(temporary_name)

File: frontend/lib/call-analytics.ts
export type CallChannel = 'browser' | 'sip';
export type CallOutcome = 'successful' | 'failed';
export type CallDetail =
  | 'grounded_answer_delivered'
  | 'tax_rule_delivered'
  | 'market_quote_delivered'
  | 'historical_return_calculated'
  | 'paper_fill_completed'
  | 'human_help_created'
  | 'no_completed_action'
  | 'incomplete'
  | 'no_response'
  | 'tool_unavailable'
  | 'system_error';

export interface CallAnalyticsSummary {
  version: 2;
  success_definition: string;
  totals: {
    total_calls: number;
    successful_calls: number;
    failed_calls: number;
    success_rate_percent: number;
    total_duration_seconds: number;
    fined_talk_seconds: number;
    taxed_talk_seconds: number;
    handoff_count: number;
  };
  recent_calls: Array<{
    call_id: string;
    started_at: string;
    duration_seconds: number;
    channel: CallChannel;
    outcome: CallOutcome;
    detail: CallDetail;
    fined_talk_seconds: number;
    taxed_talk_seconds: number;
    handoff_count: number;
  }>;
}

const SUMMARY_KEYS = ['recent_calls', 'success_definition', 'totals', 'version'] as const;
const TOTAL_KEYS = [
  'failed_calls',
  'fined_talk_seconds',
  'handoff_count',
  'success_rate_percent',
  'successful_calls',
  'taxed_talk_seconds',
  'total_calls',
  'total_duration_seconds',
] as const;
const CALL_KEYS = [
  'call_id',
  'channel',
  'detail',
  'duration_seconds',
  'fined_talk_seconds',
  'handoff_count',
  'outcome',
  'started_at',
  'taxed_talk_seconds',
] as const;
const CALL_ID = /^CALL-(?:[A-F0-9]{4}-){5}[A-F0-9]{4}$/;
const DETAILS = new Set<CallDetail>([
  'grounded_answer_delivered',
  'tax_rule_delivered',
  'market_quote_delivered',
  'historical_return_calculated',
  'paper_fill_completed',
  'human_help_created',
  'no_completed_action',
  'incomplete',
  'no_response',
  'tool_unavailable',
  'system_error',
]);
const SUCCESS_DETAILS = new Set<CallDetail>([
  'grounded_answer_delivered',
  'tax_rule_delivered',
  'market_quote_delivered',
  'historical_return_calculated',
  'paper_fill_completed',
  'human_help_created',
]);

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function hasExactKeys(value: Record<string, unknown>, keys: readonly string[]): boolean {
  const actual = Object.keys(value).sort();
  return actual.length === keys.length && actual.every((key, index) => key === keys[index]);
}

function safeCount(value: unknown): number {
  if (!Number.isSafeInteger(value) || Number(value) < 0) throw new Error('Invalid call count');
  return Number(value);
}

function sumDoesNotExceed(values: number[], total: number): boolean {
  let sum = 0;
  for (const value of values) {
    if (value > total - sum) return false;
    sum += value;
  }
  return true;
}

export function emptyCallAnalyticsSummary(): CallAnalyticsSummary {
  return {
    version: 2,
    success_definition:
      'The learner completed a verified action using grounded evidence, trusted data, a verified tax rule, paper practice or human help.',
    totals: {
      total_calls: 0,
      successful_calls: 0,
      failed_calls: 0,
      success_rate_percent: 0,
      total_duration_seconds: 0,
      fined_talk_seconds: 0,
      taxed_talk_seconds: 0,
      handoff_count: 0,
    },
    recent_calls: [],
  };
}

export function decodeCallAnalyticsSummary(value: unknown): CallAnalyticsSummary {
  if (!isRecord(value) || !hasExactKeys(value, SUMMARY_KEYS) || value.version !== 2) {
    throw new Error('Invalid analytics summary');
  }
  if (
    typeof value.success_definition !== 'string' ||
    value.success_definition.length < 10 ||
    value.success_definition.length > 360
  ) {
    throw new Error('Invalid success definition');
  }
  if (!isRecord(value.totals) || !hasExactKeys(value.totals, TOTAL_KEYS)) {
    throw new Error('Invalid analytics totals');
  }
  const totalCalls = safeCount(value.totals.total_calls);
  const successfulCalls = safeCount(value.totals.successful_calls);
  const failedCalls = safeCount(value.totals.failed_calls);
  const totalDuration = safeCount(value.totals.total_duration_seconds);
  const finedTalk = safeCount(value.totals.fined_talk_seconds);
  const taxedTalk = safeCount(value.totals.taxed_talk_seconds);
  const handoffs = safeCount(value.totals.handoff_count);
  const rate = value.totals.success_rate_percent;
  if (
    successfulCalls + failedCalls !== totalCalls ||
    typeof rate !== 'number' ||
    !Number.isFinite(rate) ||
    rate < 0 ||
    rate > 100 ||
    Math.abs(rate - (totalCalls ? Math.round((successfulCalls * 1000) / totalCalls) / 10 : 0)) >
      0.001 ||
    finedTalk + taxedTalk > totalDuration + totalCalls
  ) {
    throw new Error('Inconsistent analytics totals');
  }
  if (!Array.isArray(value.recent_calls) || value.recent_calls.length > 20) {
    throw new Error('Invalid recent calls');
  }
  const recentCalls = value.recent_calls.map((candidate) => {
    if (!isRecord(candidate) || !hasExactKeys(candidate, CALL_KEYS)) {
      throw new Error('Invalid recent call');
    }
    if (typeof candidate.call_id !== 'string' || !CALL_ID.test(candidate.call_id)) {
      throw new Error('Invalid call ID');
    }
    if (
      typeof candidate.started_at !== 'string' ||
      !Number.isFinite(Date.parse(candidate.started_at))
    ) {
      throw new Error('Invalid call time');
    }
    const duration = safeCount(candidate.duration_seconds);
    if (duration > 86_400) throw new Error('Invalid call duration');
    if (candidate.channel !== 'browser' && candidate.channel !== 'sip') {
      throw new Error('Invalid call channel');
    }
    const channel: CallChannel = candidate.channel;
    if (candidate.outcome !== 'successful' && candidate.outcome !== 'failed') {
      throw new Error('Invalid call outcome');
    }
    const outcome: CallOutcome = candidate.outcome;
    if (typeof candidate.detail !== 'string' || !DETAILS.has(candidate.detail as CallDetail)) {
      throw new Error('Invalid call detail');
    }
    const detail = candidate.detail as CallDetail;
    if ((outcome === 'successful') !== SUCCESS_DETAILS.has(detail)) {
      throw new Error('Inconsistent call outcome');
    }
    const finedTalkSeconds = safeCount(candidate.fined_talk_seconds);
    const taxedTalkSeconds = safeCount(candidate.taxed_talk_seconds);
    const handoffCount = safeCount(candidate.handoff_count);
    if (finedTalkSeconds + taxedTalkSeconds > duration + 1) {
      throw new Error('Inconsistent call speaking time');
    }
    return {
      call_id: candidate.call_id,
      started_at: candidate.started_at,
      duration_seconds: duration,
      channel,
      outcome,
      detail,
      fined_talk_seconds: finedTalkSeconds,
      taxed_talk_seconds: taxedTalkSeconds,
      handoff_count: handoffCount,
    };
  });
  const recentSuccessfulCalls = recentCalls.filter((call) => call.outcome === 'successful').length;
  const recentFailedCalls = recentCalls.length - recentSuccessfulCalls;
  if (
    recentCalls.length > totalCalls ||
    recentSuccessfulCalls > successfulCalls ||
    recentFailedCalls > failedCalls ||
    !sumDoesNotExceed(
      recentCalls.map((call) => call.duration_seconds),
      totalDuration
    ) ||
    !sumDoesNotExceed(
      recentCalls.map((call) => call.fined_talk_seconds),
      finedTalk
    ) ||
    !sumDoesNotExceed(
      recentCalls.map((call) => call.taxed_talk_seconds),
      taxedTalk
    ) ||
    !sumDoesNotExceed(
      recentCalls.map((call) => call.handoff_count),
      handoffs
    )
  ) {
    throw new Error('Recent calls exceed analytics totals');
  }

  return {
    version: 2,
    success_definition: value.success_definition,
    totals: {
      total_calls: totalCalls,
      successful_calls: successfulCalls,
      failed_calls: failedCalls,
      success_rate_percent: rate,
      total_duration_seconds: totalDuration,
      fined_talk_seconds: finedTalk,
      taxed_talk_seconds: taxedTalk,
      handoff_count: handoffs,
    },
    recent_calls: recentCalls,
  };
}
